algorithms/kernel/k_nearest_neighbors/kdtree_knn_impl.i

```diff
@@ -80,91 +80,84 @@ inline const T & max(const T & a, const T & b)
 template <typename T, CpuType cpu>
 class Stack
 {
-public:
-    Stack() : _data(nullptr) {}
-
-    ~Stack()
+private:
+    T * _data;
+    long _top;
+    long _sizeMinus1;
+    size_t _size;
+    /*
+        There was a strong assumption that stack can be initialized by only powers of 2.
+        Due to the usage scenario it's true but wrong in general. This function should fix it.
+    */
+    size_t nextPowerOf2(size_t x)
```

Comment:
Looks like a magic. Is it possible to write this simpler?

Reply:
Yes, there're some techniques. This one is just the most elegant & also performance optimized. Have I use another implementation?

Reply:
If it's simpler and doesn't drop **TOTAL** algorithm performance - yes

Reply:
As example you can use _lzcnt_u64 intrinsic or corresponding instruction.